### DISCUSSION TITLE
Why the order of `visualize_array_sharding()` is so strange?

### DISCUSSION THREAD

**Discussion by yixiaoer** [jax-ml/jax#19661] [Q&A] [answered]
The order of devices in the output of `visualize_array_sharding()` is very strange. It is in the order of 0, 1, 2, 3, 6, 7, 4, 5 instead of 0, 1, 2, 3, 4, 5, 6, 7. Why is that?
![](https://github.com/google/jax/assets/33915732/e5350b2d-544d-40b3-9255-3cb13bbe1338)
![](https://github.com/google/jax/assets/33915732/3df4063f-4e48-4c70-a0af-4b836b13fa97)
![](https://github.com/google/jax/assets/33915732/1a0cbee0-28c0-4b1b-8b61-db9ef36aa883)
![](https://github.com/google/jax/assets/33915732/b166fd00-6c59-4a3a-b498-235ee138be10)

**Comment by jakevdp** [comment, marked as answer]
Hi - thanks for the question! In auto-sharding contexts, JAX does its best to lay out data in a manner that is most efficient given the topology of the device(s). In the case of TPU, the topology may be a 3D mesh or a torus, depending on the system (look for *Interconnect Topology* in the [Cloud TPU Docs](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm)).

You can get a representation of the topology you are working with by printing the coordinates of the TPU devices:
```python
import jax
print([d.coords for d in jax.devices()])
```
Does that answer your question?

**Reply by yixiaoer** [reply]
Yes, it makes sense. Thanks for the explanation!
